### ISSUE TITLE
Data Protection(DPS) with NVMe RDMA target.

### ISSUE THREAD

**Issue by ashish19121** [spdk/spdk#830]
Please use the issue tracker only for reporting suspected issues.

See [The SPDK Community Page](http://www.spdk.io/community/) for other SPDK communications channels.

I have configured SPDK as NVMe RDMA Target. I also have NVMe PCIe device that is formatted with 520 sectors and type 3 protection(exposed via NVMe RDMA SPDK target). I am able to connect to controller and see namespace with 520 sector size and type 3 protection on the initiator.

Now I am running IO which has both dptr(data SGL) and mptr (metadata SGL). Now I want to understand how SPDK (target) handles such commands (with both dptr and mptr). Also how nvme rdma target handles such commands ?  Let me know if this is supported and if supported can you point me to the code ? I want to understand how NVMe commands with both dptr and mptr are mapped to RDMA request.

**Comment by shuhei-matsumoto** [comment]
Hi @ashish19121 

Metadata (if supported) is transferred as a contiguous part of logical blocks. Specifically,
transferring metadata as a separate buffer of data is not supported.

Hence the following feature will be necessary if we do by software.

*  the SPDK NVMe RDMA driver will have to receive data and separate metadata, verifies metadata, merges metadata with data, and then transfers extended LBA payload.

*  On the other hand, the SPDK NVMe-oF target receives interleaved data and metadata, verifies metadata, separates metadata from extended LBA payload, and transfers data and separate metadata to bdev layer.

SPDK doesn't have such features yet. Traditionally such features has been done by hardware offload in FC and iSCSI.

**Comment by ashish19121** [comment]
Are we planning to support data integrity for NVMe RDMA case ?

Thanks,
Ashish R

**Comment by shuhei-matsumoto** [comment]
Hi @ashish19121 

If the data integrity you want is software defined and what I commented, I don't know any plan in SPDK community yet. SPDK NVMe-oF target supports PI pass-through now but I think that is not what you want.

**Comment by shuhei-matsumoto** [comment]
I suppose you need high performance because your use case is NVMe-RDMA, and  hardware offload may be nice to have. Sasha and his colleagues may be working on such offload feature but they will have their own roadmap and priority.

**Comment by shuhei-matsumoto** [comment]
Hi @ashish19121 

I added two cards for this request to trello.
https://trello.com/c/s2WQ0wvA/55-implement-t10-dif-and-dix-conversion
https://trello.com/c/euE4Dnda/22-implement-t10-dif-and-dix-conversion
Please add any comment on them.

**Comment by shuhei-matsumoto** [comment]
Hi @ashish19121 

We can track such features on Trello board and so close this issue.
Please open again if any update.

Thanks,
Shuhei

**Comment by sashakot** [comment]
Hi @ashish19121

My group in Mellanox is working on public user-space API for data protection offload. This API are going to be submitted to rdma-core by October. In parallel, we're evaluating possibility  t10diff offload capabilities to NVME-OF RDMA in SPDK. 
We're thinking about the same approach like @shuhei-matsumoto  introduced in TCP transport  : "Local" mode for data protection. 

Best regards
Sasha

**Comment by shuhei-matsumoto** [comment]
Thank you for your update, @sashakot . Great news to SPDK community :)

**Comment by ashish19121** [comment]
Thanks for the update Sasha :)

**Comment by aaronknister** [comment]
Hi @sashakot, do you know when we might see the user-space API land in rdma-core? I checked and I don't think it's there.

**Comment by sashakot** [comment]
> Hi Sasha Kotchubievsky, do you know when we might see the user-space API land in rdma-core? I checked and I don't think it's there.

Hi, 

We plan to submit the "Signature offload" ("Data protection offload") to rdma-core in June 2020

Best regards
Sasha
